### PR TITLE
Retag as FnEntry on `drop_in_place`

### PR DESCRIPTION
**Pull request by JakobDegen**
This commit changes the mir drop shim to always retag its argument as if it were a `&mut`.

cc rust-lang/unsafe-code-guidelines#373

### CODE REVIEW

**Comment by JakobDegen** [comment]
Eh, sorry, forgot to r? ghost, feel free to unassign yourself

File: compiler/rustc_mir_transform/src/shim.rs [hunk 1]
```diff
@@ -176,16 +176,25 @@ fn build_drop_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, ty: Option<Ty<'tcx>>)
 
     if ty.is_some() {
         // The first argument (index 0), but add 1 for the return value.
-        let dropee_ptr = Place::from(Local::new(1 + 0));
+        let mut dropee_ptr = Place::from(Local::new(1 + 0));
         if tcx.sess.opts.unstable_opts.mir_emit_retag {
             // Function arguments should be retagged, and we make this one raw.
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/shim.rs:181]: The comment needs updating.

File: compiler/rustc_mir_transform/src/shim.rs [hunk 2]
```diff
@@ -176,16 +176,25 @@ fn build_drop_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, ty: Option<Ty<'tcx>>)
...
+            dropee_ptr = body.local_decls.push(LocalDecl::new(ref_ty, span)).into();
+            let statements = [
+                StatementKind::Assign(Box::new((dropee_ptr, reborrow))),
+                StatementKind::Retag(RetagKind::FnEntry, Box::new(dropee_ptr)),
+            ];
+            for s in statements {
+                body.basic_blocks_mut()[START_BLOCK]
+                    .statements
+                    .push(Statement { source_info, kind: s });
+            }
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/shim.rs:197]: The previous code added things at the *beginning* of the start block, you are moving it to the end. That seems strange. Why did you change that?

I also don't understand why this is so different from the old code -- why is there also an assignment being generated? This needs tons of comments to explain all that happens.

**Comment by RalfJung** [comment]
Yeah assigning ghost isn't a thing any more with triagebot, I think...

File: compiler/rustc_mir_transform/src/shim.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir_transform/src/shim.rs:197]: Oh, you need the assignment because `Retag` ignores raw pointers. Makes sense but needs a comment.

But the point about inserting at the beginning vs the end remains.

Comment (JakobDegen) [reply compiler/rustc_mir_transform/src/shim.rs:197]: There are no statements in the body at this time, so the beginning and end are the same.

I will add comments and more tests once we actually decide if we want this behavior, this PR as is was not meant to be mergable

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/shim.rs:197]: > There are no statements in the body at this time, so the beginning and end are the same.

Then please `assert!` that. I added them at the beginning just to be sure.

Comment (JakobDegen) [reply compiler/rustc_mir_transform/src/shim.rs:204]: I've added a comment to this, but the body is created like 10 lines up, so I don't really think it's worth an assert

**Comment by JakobDegen** [comment]
@saethlin had tested this and found no regressions in the ecosystem, so seems reasonable to have. Marking as ready:

r? @RalfJung 

The MIR gen for this is covered by an existing test. I'd also like to add a test to ensure that we find the UB in Miri - should I be adding that to the miri subtree in this PR or submitting a separate PR to rust-lang/miri?

**Comment by RalfJung** [comment]
Please add the miri test in the same PR. It's great that we can finally do this. :)

File: compiler/rustc_mir_transform/src/shim.rs [hunk 3]
```diff
@@ -176,16 +176,32 @@ fn build_drop_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, ty: Option<Ty<'tcx>>)
...
-                    kind: StatementKind::Retag(RetagKind::Raw, Box::new(dropee_ptr)),
-                },
+            // We want to treat the function argument as if it was passed by `&mut`. As such, we
+            // generate
+            // ```
+            // temp = &mut *arg;
+            // Retag(temp, FnEntry)
+            // ```
+            // It's important that we do this first, before anything that depends on what `dropee_ptr`
+            // is has been put into the body.
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/shim.rs:188]:
```suggestion
            // has been put into the body.
```

**Review by RalfJung — COMMENTED** [review]
LGTM, apart from the missing test.

**Comment by RalfJung** [comment]
In fact we probably want at least 2 tests -- one for the aliasing concerns and one for the validity (aligned and non-null).

**Comment by RalfJung** [comment]
@rustbot author
@JakobDegen let me know if you need help adding the test. :)

**Comment by JakobDegen** [comment]
Sorry for the delay, was traveling. Tests added. @rustbot ready

File: src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#[repr(transparent)]
+struct HasDrop(u8);
```
Comment (RalfJung) [inline src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs:2]: This test already [has UB with current Miri](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8bce42233ba3b7b16349cdf14ab86204), so it doesn't really test your changes.

Comment (RalfJung) [reply src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs:2]: [Here is a version](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5c230abdc45b4c55d4cdca8233cd7c97) that currently passes, which hopefully is declared UB with your patch.

Comment (RalfJung) [reply src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs:2]: And [here is another test](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=483452b2a2066b6e15f50c7eeecf9c25).

Comment (JakobDegen) [reply src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs:2]: The semantics we have right now are that `drop_in_place` is still a nop if the type genuinely has no drop glue at all. I've adjusted the second test accordingly.

Comment (RalfJung) [reply src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs]: Oh, is that compatible with @pcwalton's plans? And why is that, can't we make it always retag?

The second test was quite deliberately designed to hit the case of a type that does not need dropping.

Comment (JakobDegen) [reply src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs]: Yes, it is, since we never codegen nop `drop_in_place` (either for direct function calls or `TerminatorKind::Drop`). This is the result of the check [here](https://github.com/rust-lang/rust/blob/07e664c41ae5679b912ba9d1f7364ebaac8e51b6/compiler/rustc_ty_utils/src/instance.rs#L60-L78). Although codegen also separately [ignores these](https://github.com/rust-lang/rust/blob/1eb62b1235fd77200e6bd967d70e83c0f2497233/compiler/rustc_codegen_ssa/src/mir/block.rs#L768-L772). That should make sure everything is compatible, but I'll add a test ensuring it stays that way

Comment (RalfJung) [reply src/tools/miri/tests/fail/stacked_borrows/drop_in_place.rs]: > and as best I can tell from reading the LLVM docs putting noalias on an unused argument probably does nothing (although who knows).

There's still `dereferenceable aligned nonnull` though, right?

**Comment by RalfJung** [comment]
Btw, force-pushing makes it very hard to review just the changes since the last time I looked at things. Please only force-push when necessary due to a rebase, or at the end after review is done to clean up the history.

File: src/tools/miri/tests/pass/drop_in_place_null.rs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// Make sure that dropping types with no drop glue is DB even for invalid pointers.
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/drop_in_place_null.rs:1]: This is totally against what we [document](https://doc.rust-lang.org/nightly/core/ptr/fn.drop_in_place.html#safety). So I think we should make this UB. If that's a problem, clearly the documentation must be updated.

**Comment by JakobDegen** [comment]
The additional commit changes this to unconditionally retag. @saethlin could you rerun this? It's quite possible that this introduces new UB.

This change also affects both explicit `drop_in_place` and `Drop` terminators. I'm fairly sure that is not what we want and we'll actually want to change that, although it's *possible* that it's not observable in MIR that we create today (I'll be able to demonstrate it once I add some additional things to custom mir though)

**Comment by RalfJung** [comment]
> This change also affects both explicit drop_in_place and Drop terminators. I'm fairly sure that is not what we want and we'll actually want to change that, 

Why?

File: src/tools/miri/tests/fail/stacked_borrows/drop_in_place_retag.rs [hunk 6]
```diff
@@ -3,16 +3,9 @@
```
Comment (RalfJung) [inline src/tools/miri/tests/fail/stacked_borrows/drop_in_place_retag.rs:3]: Might be worth indicating in the test name and/or comment that this is specifically about `drop_in_place` on a type that does not need dropping.

**Comment by JakobDegen** [comment]
> Why?

Imagine for the sake of argument that we offered users a way to directly call `RetagFnEntry` as an intrinsic of some kind. I'd expect the following to happen:

```rust
fn non_generic(x: u8) {
    let r = &mut x;
    RetagFnEntry(r);
    // No implicit drop terminator because `u8` is copy
}

fn generic<T>(x: T) {
    let r = &mut x;
    RetagFnEntry(r);
    // Unconditional UB here, because the retag for the drop
    // would attempt to pop the tag of `r`
}

fn main() {
    non_generic(0_u8); // DB
    generic(0_u8); // UB
}
```

This seems not great. More generally, I expect that rustc is today and will always be incredibly inconsistent about where it inserts `Drop`s for types that do not have drop glue. It seems much better for everyone if that doesn't matter.

**Comment by RalfJung** [comment]
Hm. Note that this use of FnEntry is invalid since FnEntry should only ever be done right at the start of a function. But I take the wider point about `Drop` terminators on non-drop types.

`Drop` terminators are not the same as regular function calls anyway, though. Does codegen even emit a call to `drop_in_place` when the (monomorphized) type is non-dropping? If not, we could easily have the same logic in Miri.

**Comment by bjorn3** [comment]
> Does codegen even emit a call to drop_in_place when the (monomorphized) type is non-dropping?

No

**Comment by JakobDegen** [comment]
> If not, we could easily have the same logic in Miri.

Yes, this is what I'm suggesting.

**Comment by RalfJung** [comment]
Yeah that should be easy to do here

https://github.com/rust-lang/rust/blob/9e772114e6a17498f1ba27ba96693012788972e2/compiler/rustc_const_eval/src/interpret/terminator.rs#L122-L129

Do you want to do that in this PR?

**Comment by saethlin** [comment]
Get back to you on what? I wasn't aware you were waiting for something from me.

**Comment by saethlin** [comment]
Will do

**Comment by RalfJung** [comment]
> Yeah, I can do that here later. Would prefer to wait for Ben to get back to us first though, to see how much (if any) code violates this.

I worry this might take a bit long, and we are blocking https://github.com/rust-lang/rust/pull/103614 on this. It seems very unlikely that someone would call `drop_in_place` on a pointer that they don't have write permissions for.

**Comment by RalfJung** [comment]
@rustbot author
(I think this reflects the current state? https://github.com/rust-lang/rust/pull/103957#issuecomment-1339139960 is open.)

**Comment by saethlin** [comment]
I agree this whole thing is unlikely to turn up anything.

> I worry this might take a bit long

Well, I back-burnered this because of the Bad Time I have had with building a cargo-miri that can run doctests out of a rust branch, which is probably the bootstrap problem "stage numbering for tools" lightly discussed on Zulip in t-infra/bootstrap.

If anyone wants to do this regression experiment themselves, all you need to do is change https://github.com/saethlin/miri-tools/blob/main/docker/Dockerfile to build this branch, `cargo r -- --crates=10000` or such, find the files under `logs/` that contain `Undefined Behavior`, and put those in a file with lines like `name==version`, `mv` all the logs somewhere else, then revert the Dockerfile changes and `cargo r --crate-list=my-crates-file`, and investigate anything that doesn't report UB.

If someone can show me a branch of my repo that builds this branch and can run doctests, I'm happy to donate my CPU cycles, but I find fighting with the rustc build system in a Docker container exceedingly frustrating and I've lots of less-frustrating things to work on lately.

**Comment by JakobDegen** [comment]
Rebased and attached a commit to the end. Drop terminators are now a complete nop if the type has no drop glue and are documented as such.

@rustbot ready

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 7]
```diff
@@ -562,14 +562,13 @@ pub enum TerminatorKind<'tcx> {
...
-    /// glue is special - this would be part of Rust's memory model. (**FIXME**: due we have an
-    /// issue tracking if drop glue has any interesting semantics in addition to those of a function
-    /// call?)
-    ///
-    /// `Drop` before drop elaboration is a *conditional* execution of the drop glue. Specifically, the
-    /// `Drop` will be executed if...
+    ///
+    /// After drop elaboration: `Drop` terminators are a complete nop for types that have no drop
+    /// glue. For other types, `Drop` terminators behave exactly like a call to
+    /// `core::mem::drop_in_place` with a pointer to the given place.
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/syntax.rs:568]: This is the key change in the MIR spec here. It reflects what the codegen backends already do.
